pkgs/by-name/xa/xar/patches/0020-Fall-back-to-readlink-on-Linux.patch

```diff
@@ -0,0 +1,57 @@
+From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
+From: Randy Eckenrode <randy@largeandhighquality.com>
+Date: Wed, 30 Oct 2024 20:19:20 -0400
+Subject: [PATCH 20/20] Fall back to readlink on Linux
+
+---
+ xar/lib/archive.c | 23 +++++++++++++++++++++++
+ 1 file changed, 23 insertions(+)
+
+diff --git a/xar/lib/archive.c b/xar/lib/archive.c
+index b7f9cbf..85567f7 100644
+--- a/xar/lib/archive.c
++++ b/xar/lib/archive.c
+@@ -95,6 +95,8 @@ void xar_serialize(xar_t x, const char *file);
+ static pthread_mutex_t xar_new_mutex = PTHREAD_MUTEX_INITIALIZER;
+ #endif // HAVE_LIBPTHREAD
+ 
++#define NUM_DIGITS(n) (n == 0 ? 0 : 1 + NUM_DIGITS(n / 10))
++
+ /* xar_new
+  * Returns: newly allocated xar_t structure
+  * Summary: just does basicallocation and initialization of 
+@@ -507,10 +509,31 @@ xar_t xar_fdopen_digest_verify(int fd, int32_t flags, void *expected_toc_digest,
+ 	// If there are hardlinks, the path we pick is the most recently opened by
+ 	// the filesystem; which is effectively random.
+ 	char path_buff[PATH_MAX];
++#if defined(__APPLE__)
+ 	if (fcntl(fd, F_GETPATH, path_buff) < 0) {
+ 		close(fd);
+ 		return NULL;
+ 	}
++#elif defined(__linux__)
++	// Linux has to get the path to the file via `/proc`.
++	char proc_buff[PATH_MAX];
++	if (snprintf(proc_buff, PATH_MAX, "/proc/self/fd/%i", fd) < 0) {
++		close(fd);
++		return NULL;
++	}
++	if (readlink(proc_buff, &path_buff, PATH_MAX) < 0) {
++		close(fd);
++		return NULL;
++	}
++	// The filename is the file’s when the fd was created. Check to make sure it still exists.
```

Comment:
Do we really have to check that the file exists?

Reply:
Not sure. The documentation just says it gets the path. I’m going to write some test programs later today to probe the behavior to see what Darwin does.

Reply:
Darwin tracks name changes for renames. When the file is unlinked, `F_GETPATH` appears to return the last name it had. It may be possible to track renames in Linux (not sure), but it’s probably not worth the effort. Bailing if the underlying file was removed or renamed seems like the safest/easiest option.